lib/Parse/ParseDecl.cpp

```diff
@@ -3256,9 +3256,17 @@ void Parser::parseDeclDelayed() {
       } else if (auto *ED = dyn_cast<ExtensionDecl>(parent)) {
         ED->addMember(D);
       } else if (auto *SF = dyn_cast<SourceFile>(parent)) {
-        // FIXME: unify notification to ASTSourceFileScope with addMember
-        // mechanism used above
         SF->Decls.push_back(D);
+      } else if (auto *CE = dyn_cast<ClosureExpr>(parent)) {
+        // Replace the closure body with one including the new Decl.
+        auto *body = CE->getBody();
+        SmallVector<ASTNode, 8> Elts(body->getElements().begin(),
+                                     body->getElements().end());
+        Elts.push_back(ASTNode(D));
```

Comment:
Thanks. I've added ordering to this code. What are you thinking w.r.t. single-expression closures? As far as when does this come up, I can run the tests again with an abort in there and let you know.

Reply:
Honestly, I'm not sure if this really makes sense either. I think we just can't delay something out of a BraceStmt today, whether from a closure or not, and we shouldn't bother writing code to support it.

Reply:
I inserted a forced assertion failure into this code and ran `ninja check-swift-validation` to see what happened. In the log file, my inserted assertion was triggered 8 times in:

- IDE/crashers_fixed/030-swift-mangle-mangler-mangledecltype.swift
- IDE/crashers_fixed/029-swift-decl-walk.swift
- IDE/crashers_fixed/032-swift-expr-propagatelvalueaccesskind.swift
- IDE/crashers_fixed/047-swift-typechecker-typechecktoplevelcodedecl.swift
- IDE/crashers_fixed/080-swift-valuedecl-getinterfacetype.swift
- IDE/crashers_fixed/084-swift-parser-consumedecl.swift
- IDE/complete_crashes.swift
- IDE/complete_in_closures.swift

In each case, it was running `swift-ide-test`.

So, it does happen in those contexts.

However, in each case, it's doing code completion, so this code would not be required until ASTScope lookup handles completion completely. I'm planning to land ASTScopeLookup with it initially disabled-by-default when completion is being done. So, I don't need this code right way. 

Hmmm... What I can do is take it out of this PR, and put it in a radar for me to come back to in the future. That's probably the best course for now.

Assuming that you're surprised that this actually happens, join the club! I and I think @slavapestov  and @DougGregor  have also been surprised at some of the things going on with our AST!

Thanks for your help.

Reply:
I suppose I'm not surprised it happens for completion, but I don't know what we're going to do about completion! Still, maybe it can be appended to the end but with an assertion that that's the right place for it. Anyway, this looks good without it for now.

Reply:
Thanks!